src/spatial/nearest_neighbor.cpp

```diff
@@ -0,0 +1,281 @@
+
+public:
+    using DistanceType = igraph_real_t;
+
+    GraphBuildingResultSet(const igraph_integer_t neighbors, const igraph_real_t distance) :
+        max_distance(distance),
+        max_neighbors(neighbors) {}
+
+    bool addPoint(const igraph_real_t distance, const igraph_integer_t index) {
+        igraph_integer_t i;
+
+        if (index == current_vertex) {
+            return true;
+        }
+
+        for (i = current_added; i > 0; i--) {
+            if ((dists[i-1] > distance /*|| dists[i-1] == distance && index < edges[i-1]*/ )) {
+                    if (i < max_neighbors) {
+                    dists[i] = dists[i-1];
+                    edges[i] = edges[i-1];
+                }
+            } else {
+                break;
+            }
+        }
+        if (i < max_neighbors) {
+            edges[i] = index;
+            dists[i] = distance;
+        }
+        if (current_added != max_neighbors) {
+            current_added++;
+        }
+        return true;
+    }
+
+    void init(igraph_real_t *dists_, igraph_integer_t *edges_, igraph_integer_t current_vertex_) {
+        edges = edges_;
+        dists = dists_;
+        current_added = 0;
+        current_vertex = current_vertex_;
+    }
+
+    void sort() { }
+
+    igraph_integer_t size() const {
+        return current_added;
+    }
+
+    bool full() const {
+        return current_added == max_neighbors;
+    }
+
+    bool empty() const {
+        return current_added == 0;
+    }
+
+    igraph_real_t worstDist() const {
+        if (current_added < max_neighbors || current_added == 0) {
+            return  max_distance;
+        }
+        return dists[current_added-1];
+    }
+
+};
+
+template <typename Metric, igraph_integer_t Dimension>
+static igraph_error_t neighbor_helper(
+        igraph_t *graph,
+        const igraph_matrix_t *points,
+        igraph_integer_t neighbors,
+        igraph_real_t cutoff,
+        igraph_integer_t dimension,
+        igraph_bool_t directed) {
+
+    const igraph_integer_t point_count = igraph_matrix_nrow(points);
+
+    using kdTree = nanoflann::KDTreeSingleIndexAdaptor<Metric, ig_point_adaptor, Dimension>;
+
+    ig_point_adaptor adaptor(points);
+
+    kdTree tree(dimension, adaptor, nanoflann::KDTreeSingleIndexAdaptorParams(10));
+
+    tree.buildIndex();
+
+    igraph_vector_t current_point;
+    IGRAPH_VECTOR_INIT_FINALLY(&current_point, dimension);
+
+    igraph_integer_t neighbor_count = neighbors >= 0 ? neighbors : point_count;
+
+    GraphBuildingResultSet results(neighbor_count, cutoff);
+    std::vector<igraph_integer_t> neighbor_set(neighbor_count);
+    std::vector<igraph_real_t>    distances(neighbor_count);
+    std::vector<igraph_integer_t> edges;
+    for (igraph_integer_t i = 0; i < point_count; i++) {
+
+        std::fill(neighbor_set.begin(), neighbor_set.end(), 0);
+        std::fill(distances.begin(), distances.end(), 0);
+
+        results.init(distances.data(), neighbor_set.data(), i);
```

Comment:
`distances` and `neighbor_set` are accessed in the result class through their `.data()` pointer. Since we're writing C++, is there any reason you don't wan to access them in the standard way as `std::vector`s, using `[]`?